File: 3418-maximum-amount-of-money-robot-can-earn/3418-maximum-amount-of-money-robot-can-earn.java
class Solution {
    int n, m;
    int[][] coins;
    int[][][] memo;

    public int maximumAmount(int[][] coins) {
        this.coins = coins;
        n = coins.length;
        m = coins[0].length;

        // 3D memo: n x m x 3
        memo = new int[n][m][3];

        // initialize with a sentinel (not visited)
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                for (int k = 0; k < 3; k++) {
                    memo[i][j][k] = Integer.MIN_VALUE;
                }
            }
        }

        return helper(0, 0, 2);
    }

    private int helper(int r, int c, int k) {
        // out of bounds
        if (r >= n || c >= m) {
            return Integer.MIN_VALUE / 2; // avoid overflow
        }

        // memo check
        if (memo[r][c][k] != Integer.MIN_VALUE) {
            return memo[r][c][k];
        }

        int value = coins[r][c];

        // destination
        if (r == n - 1 && c == m - 1) {
            if (value < 0 && k > 0) {
                return memo[r][c][k] = 0;
            }
            return memo[r][c][k] = value;
        }

        // option 1: take value
        int take = value + Math.max(
            helper(r, c + 1, k),
            helper(r + 1, c, k)
        );

        // option 2: neutralize robber
        int skip = Integer.MIN_VALUE;
        if (value < 0 && k > 0) {
            skip = Math.max(
                helper(r, c + 1, k - 1),
                helper(r + 1, c, k - 1)
            );
        }

        memo[r][c][k] = Math.max(take, skip);
        return memo[r][c][k];
    }
}

File: 3418-maximum-amount-of-money-robot-can-earn/3418-maximum-amount-of-money-robot-can-earn.py
class Solution:
    def maximumAmount(self, coins):
        n, m = len(coins), len(coins[0])

        # dp[r][c][k] = max coins from (r,c) to end
        dp = [[[float('-inf')] * 3 for _ in range(m)] for _ in range(n)]

        # 🎯 Fill from bottom-right → top-left
        for r in range(n - 1, -1, -1):
            for c in range(m - 1, -1, -1):
                for k in range(3):

                    val = coins[r][c]

                    # 🟢 Base case (destination)
                    if r == n - 1 and c == m - 1:
                        if val < 0 and k > 0:
                            dp[r][c][k] = 0
                        else:
                            dp[r][c][k] = val
                        continue

                    # next positions
                    down = dp[r+1][c][k] if r + 1 < n else float('-inf')
                    right = dp[r][c+1][k] if c + 1 < m else float('-inf')

                    # ✅ option 1: take value
                    take = val + max(down, right)

                    # ✅ option 2: neutralize
                    skip = float('-inf')
                    if val < 0 and k > 0:
                        down2 = dp[r+1][c][k-1] if r + 1 < n else float('-inf')
                        right2 = dp[r][c+1][k-1] if c + 1 < m else float('-inf')
                        skip = max(down2, right2)

                    dp[r][c][k] = max(take, skip)

        return dp[0][0][2]